### PR TITLE
Streamline C heap allocations

### PR DESCRIPTION
**Pull request by jkotas**
- Remove unnecessary layers in heap allocation APIs
- Delete HeapCreate/Alloc/Free/Destroy from the PAL

### CODE REVIEW

File: src/coreclr/src/ToolBox/superpmi/superpmi-shared/compileresult.cpp [hunk 1]
```diff
@@ -43,11 +41,6 @@ CompileResult::~CompileResult()
 
     if (CallTargetTypes != nullptr)
         delete CallTargetTypes;
-
-#ifndef TARGET_UNIX // PAL doesn't have HeapDestroy()
-    if (codeHeap != nullptr)
```
Comment (jkotas) [inline src/coreclr/src/ToolBox/superpmi/superpmi-shared/compileresult.cpp]: @dotnet/jit-contrib superpmi allocated a special heap for code on Windows, but not on Unix. I have unified it to not use the special heap everywhere. Any concerns about this?

File: src/coreclr/src/utilcode/clrhost_nodependencies.cpp [hunk 2]
```diff
@@ -69,26 +69,19 @@ thread_local ClrDebugState* t_pClrDebugState;
 // Fls callback to deallocate ClrDebugState when our FLS block goes away.
 void FreeClrDebugState(LPVOID pTlsData)
 {
-#ifdef _DEBUG
```
Comment (janvorli) [inline src/coreclr/src/utilcode/clrhost_nodependencies.cpp]: Why do we turn this on for Release builds now?

File: src/coreclr/src/debug/ee/debugger.cpp [hunk 3]
```diff
@@ -16763,32 +16752,25 @@ void *DebuggerHeap::Alloc(DWORD size)
...
+#ifdef HOST_WINDOWS
+    HANDLE hExecutableHeap hExecutableHeap  = ClrGetProcessExecutableHeap();
 
-#ifdef TARGET_UNIX
+    if (hExecutableHeap == NULL)
+    {
+        return NULL;
+    }
+
+    ret = ClrHeapAlloc(hExecutableHeap, 0, size);
```
Comment (janvorli) [inline src/coreclr/src/debug/ee/debugger.cpp]: Should this be just HeapAlloc?

File: src/coreclr/src/utilcode/clrhost_nodependencies.cpp (hunk 2 above)
Comment (jkotas) [reply src/coreclr/src/utilcode/clrhost_nodependencies.cpp]: This ifdef was redundant. This whole function is debug-only (ifdef a few lines above).

File: src/coreclr/src/debug/ee/debugger.cpp (hunk 3 above)
Comment (jkotas) [reply src/coreclr/src/debug/ee/debugger.cpp]: Fixed.

File: src/coreclr/src/debug/ee/debugger.cpp [hunk 4]
```diff
@@ -16506,28 +16506,21 @@ void* DebuggerHeapExecutableMemoryAllocator::Allocate(DWORD numberOfBytes)
...
     CrstHolder execMemAllocCrstHolder(&m_execMemAllocMutex);
 
     DebuggerHeapExecutableMemoryPage *pageToFreeIn = static_cast<DebuggerHeapExecutableMemoryChunk*>(addr)->data.startOfPage;
-
-    if (pageToFreeIn == NULL)
-    {
-        ASSERT(!"Couldn't locate page in which to free!");
-        return -1;
-    }
+    _ASSERTE(pageToFreeIn != NULL);
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/src/debug/ee/debugger.cpp]: This should never happen, unless we have a bug. 

If somebody passes a bad pointer into this method, we are most likely going to read a bogus pageToFreeIn and crash with bad AV. Trying to check for `pageToFreeIn == NULL` won't help in 99% of case.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**